esp/services/ws_config2/ws_config2Service.cpp

```diff
@@ -799,62 +799,67 @@ void Cws_config2Ex::getAttributes(const std::shared_ptr<EnvironmentNode> &pEnvNo
-
-                    //
-                    // Add dependencies
-                    if ((*valueIt).hasDependencies())
+                    IArrayOf<IEspChoiceType> choices;
+                    for (auto valueIt=allowedValues.begin(); valueIt!=allowedValues.end(); ++valueIt)
                     {
-                        IArrayOf<IEspDependentValueType> dependencies;
-                        for (auto &depIt: (*valueIt).getDependencies())
+                        Owned<IEspChoiceType> pChoice = createChoiceType();
+                        pChoice->setDisplayName((*valueIt).m_displayName.c_str());
+                        pChoice->setValue((*valueIt).m_value.c_str());
+                        pChoice->setDesc((*valueIt).m_description.c_str());
+                        pChoice->setMsg((*valueIt).m_userMessage.c_str());
+                        pChoice->setMsgType((*valueIt).m_userMessageType.c_str());
+
+                        //
+                        // Add dependencies
+                        if ((*valueIt).hasDependencies())
                         {
-                            Owned<IEspDependentValueType> pDep = createDependentValueType();
-                            pDep->setAttributeName(depIt.m_attribute.c_str());
-                            pDep->setAttributeValue(depIt.m_value.c_str());
-                            dependencies.append(*pDep.getLink());
+                            IArrayOf<IEspDependentValueType> dependencies;
+                            for (auto &depIt: (*valueIt).getDependencies())
+                            {
+                                Owned<IEspDependentValueType> pDep = createDependentValueType();
+                                pDep->setAttributeName(depIt.m_attribute.c_str());
+                                pDep->setAttributeValue(depIt.m_value.c_str());
+                                dependencies.append(*pDep.getLink());
+                            }
+                            pChoice->setDependencies(dependencies);
                         }
-                        pChoice->setDependencies(dependencies);
-                    }
 
-                    //
-                    // Add optional/required attributes.
-                    if (!(*valueIt).m_optionalAttributes.empty())
-                    {
-                        StringArray attributeNames;
-                        for (auto &attr: (*valueIt).m_optionalAttributes)
+                        //
+                        // Add optional/required attributes.
+                        if (!(*valueIt).m_optionalAttributes.empty())
                         {
-                            StringBuffer atrrname(attr.c_str());
-                            attributeNames.append(atrrname);
+                            StringArray attributeNames;
+                            for (auto &attr: (*valueIt).m_optionalAttributes)
+                            {
+                                StringBuffer atrrname(attr.c_str());
```

Comment:
It might be worthwhile to declare this StringBuffer outside the inner loop and either atrrname.set(attr.c_str()) or atrrname.clear().append(attr.c_str());

Reply:
Change made. Incidentally I have used this style for a while so that if I ever wanted to move the block to a method, simply selecting the block of code brings everything I need, including local variables.